src/main/java/net/minecraftforge/client/SkyRenderer.java

```diff
@@ -0,0 +1,490 @@
+            else handlers.remove(id);
+        }
+
+        boolean enabled() {
+            return !handlers.equals(this.defHandlers);
+        }
+
+        void render(float partialTicks, WorldClient world, Minecraft mc)
+        {
+            for(RenderHandler handler : handlers.values())
+            {
+                handler.handler.render(partialTicks, world, mc);
+            }
+        }
+    }
+
+    private static class RenderHandler implements Comparable<RenderHandler> {
+        IRenderHandler handler;
+        double priority;
+
+        RenderHandler(IRenderHandler handler, double priority)
+        {
+            this.handler = handler;
+            this.priority = priority;
+        }
+
+        @Override
+        public int compareTo(RenderHandler o)
+        {
+            // Higher priority comes first
+            return -Comparator.<RenderHandler>comparingDouble(k -> k.priority).compare(this, o);
+        }
+    }
+
+    private static RenderHandler handler(IRenderHandler renderer, double priority) {
+        return new RenderHandler(renderer, priority);
+    }
+
+    private Map<ResourceLocation, RenderHandler> getSkyBackDefault()
+    {
+        ImmutableMap.Builder<ResourceLocation, RenderHandler> builder = ImmutableMap.builder();
+        builder.put(new ResourceLocation("sky"), handler(new SkyBackRenderer(), 0.0));
+        return builder.build();
+    }
+
+    private Map<ResourceLocation, RenderHandler> getSunMoonStarsDefault()
+    {
+        ImmutableMap.Builder<ResourceLocation, RenderHandler> builder = ImmutableMap.builder();
+        builder.put(new ResourceLocation("sun"), handler(new SunRenderer(), 2.0));
+        builder.put(new ResourceLocation("moon"), handler(new MoonRenderer(), 1.0));
+        builder.put(new ResourceLocation("stars"), handler(new StarsRenderer(), 0.0));
+        return builder.build();
+    }
+
+    private Map<ResourceLocation, RenderHandler> getSkyFrontDefault()
+    {
+        ImmutableMap.Builder<ResourceLocation, RenderHandler> builder = ImmutableMap.builder();
+        builder.put(new ResourceLocation("sky"), handler(new SkyFrontRenderer(), 0.0));
+        return builder.build();
+    }
+
+    private static final ResourceLocation MOON_PHASES_TEXTURES = new ResourceLocation("textures/environment/moon_phases.png");
+    private static final ResourceLocation SUN_TEXTURES = new ResourceLocation("textures/environment/sun.png");
+    private static final ResourceLocation END_SKY_TEXTURES = new ResourceLocation("textures/environment/end_sky.png");
+
+    private boolean vboEnabled;
+    private int glSkyList, glSkyList2;
+    private VertexBuffer skyVBO, sky2VBO;
+    private int starGLCallList;
+    private VertexBuffer starVBO;
+
+    private class SkyBackRenderer extends IRenderHandler {
+        @Override
+        public void render(float partialTicks, WorldClient world, Minecraft mc)
+        {
+            if (mc.world.provider.getDimensionType().getId() == 1)
```

Comment:
These copied renderers could be hard to port if functionality changes in the original renderer. Would it be possible to call the vanilla renderer still, but filter what is rendered depending on the current pass?

Reply:
Yes, it's possible. But I'll postpone it till it's approved by modders doing sky rendering work. I guess this whole PR is moot before that.